### ISSUE TITLE
When Che automatically opens a VS Code workspace, it does not install the recommended extensions

### ISSUE THREAD

**Issue by cgruver** [eclipse/che#22289]
### Describe the bug

When using an environment var in the devfile to set a default VS Code workspace, the recommended extensions listed in the code-workspace file are not automatically installed.

`VSCODE_DEFAULT_WORKSPACE`

If this environment var is omitted, then the extensions will be installed when the workspace is opened.

### Che version

7.68@latest

### Steps to reproduce

1. Create a workspace from: [https://github.com/cgruver/che-podman-demo.git](https://github.com/cgruver/che-podman-demo.git)
1. Open the VS Code workspace
1. Observe that the recommended extensions are installed after the workspace opens.  __Note:__ You will see the color theme change.

1. Create a workspace from: [https://github.com/eclipse-che-demo-app/che-demo-app.git](https://github.com/eclipse-che-demo-app/che-demo-app.git)
1. Observe that the included VSCode workspace is automatically opened.
1. Observe that the recommended extensions are not installed.

### Expected behavior

Extensions listed in the `code-workspace` file should be automatically installed to ease developer onboarding to a project.

```json
{
	"folders": [
		{
			"path": "."
		},
		{
			"path": "../che-demo-app-ui"
		},
		{
			"path": "../che-demo-app-service"
		}
	],
	"extensions": {
		"recommendations": [
			"redhat.vscode-yaml",
			"redhat.java",
			"redhat.vscode-quarkus",
			"redhat.vscode-openshift-connector",
			"mhutchie.git-graph",
			"eamodio.gitlens",
			"streetsidesoftware.code-spell-checker",
			"johnpapa.winteriscoming",
			"cnshenj.vscode-task-manager",
			"angular.ng-template",
			"johnpapa.angular2",
			"SonatypeCommunity.vscode-iq-plugin"
		]
	},
	"settings": {
		"workbench.colorTheme": "Winter is Coming (Light)",
		"workbench.preferredDarkColorTheme": "Winter is Coming (Dark Blue)",
		"workbench.preferredLightColorTheme": "Winter is Coming (Light)"
	}
}
```

### Runtime

OpenShift

### Installation method

OperatorHub

### Environment

macOS

### Additional context

Perhaps this feature needs to be a toggle that can control whether or not Che users want extensions to be auto installed.

**Comment by l0rd** [comment]
@cgruver thank you for reporting this. Can you please check in VS Code settings if the auto installation of recommendation is enabled? (`extensions.autoInstallRecommendations`)

**Comment by azatsarynnyy** [comment]
@cgruver could you verify it in your browser's Incognito mode, please?

**Comment by cgruver** [comment]
Hey, this actually appears to be fixed in `7.69`!

I just tried it with a new install, and everything works as expected.

I have an older cluster running that was upgraded from `7.68` to `7.69`.  I'll repeat the test there.

**Comment by cgruver** [comment]
@azatsarynnyy 

The issue does not appear in a new Che `7.69` install.

The issue still exists in another OpenShift cluster where Che was updated from `7.68` to `7.69`.

However, the issue does not appear in an incognito window as you asked me to test.

I assume this implies an issue with something cached in the browser that broke during the update from `7.67` to `7.68`?

**Comment by cgruver** [comment]
![Screenshot 2023-06-29 at 10 44 58 AM](https://github.com/eclipse/che/assets/39659182/059dee52-425a-4e5b-926f-d4f9da35cf1c)

All recommended extensions were installed, and all of the code repositories in the workspace were opened.

Looks good!

**Comment by cgruver** [comment]
@azatsarynnyy I can re-create this issue by creating a workspace, deleting the workspace, and creating it again.

Something is apparently cached in the browser that is causing VS Code not to install the recommended extensions when the workspace is created the second time.

If I clear the browser cache and reopen the running workspace, then the extensions are installed as expected.

**Comment by azatsarynnyy** [comment]
Thanks @cgruver for checking. It should be caused by the same problem as https://github.com/eclipse/che/issues/22338.
I described in short in [this comment](https://github.com/eclipse/che/issues/22338#issuecomment-1618260015).

> I have to clear the whole browser cache

To work this problem around, it's enough to clear the browser's IndexedDB. For example, for Chrome:

![image](https://github.com/eclipse/che/assets/1636395/b1e7701a-fab8-471b-b488-b9efc67aba2c)

**Comment by vitaliy-guliy** [comment]
There was a bug in VS Code https://github.com/microsoft/vscode/issues/186867, which has been fixed by https://github.com/microsoft/vscode/pull/186983

How to properly reproduce the bug with older (7.70.0) version of CheCode.
1. Open the dashboard
2. Open browser development tools ( Ctrl+Shift+I or F12 )
3. Navigate to Applications, in the tree on the left go to Stogare ->IndexedDB and  delete all the databases
4. Create a workspace, install recommended extensions
5. Refresh IndexedDB, expand database `vscode-web-state-xxxxx`, ensure there is no key `extensionsAssistant/workspaceRecommendationsIgnore`
Before fixing the bug, it was set to `true` each time after installing the recommendations
6. Delete the workspace, create the same workspace.

Extensions must be installed.

Btw, I have a feeling che-code 7.72.0 has another bug related to the recommendations (a good candidate for Hack'n'hustle to contribute to vscode)

I opened workspace settings ( File -> Preferences -> Settings -> Workspace ) and disabled autoinstalling the recommendations ( Features -> Extensions :: Auto Install Recommendations )
Then I refreshed the browser database and did not find an item `extensionsAssistant/workspaceRecommendationsIgnore` set to `false`.

**Comment by l0rd** [comment]
@azatsarynnyy @vitaliy-guliy can we close this if that has been adressed? 

What about https://github.com/eclipse/che/issues/22338? Is that fixed too?

**Comment by l0rd** [comment]
Related to [downstream issue](https://issues.redhat.com/browse/CRW-4790)

**Comment by azatsarynnyy** [comment]
**TL;DR**
For the latest released Che version which is 7.73 today, this problem should still be relevant when both conditions are met:
1. A running Che Workspace name **has been already used before**, on a particular Che instance, for another Workspace.
2. Che-Code version which was running in the previous (deleted) Workspace was older than 7.71.

The solution is to fix https://github.com/eclipse/che/issues/22338 by respecting DevWorkspace ID when generating the VS Code Workspace ID here https://github.com/che-incubator/che-code/pull/148/files.

**Details**
[The upstream VS Code patch Vitaliy has mentioned above](https://github.com/microsoft/vscode/pull/186983) indirectly affects the Che-Code's problem with the recommendations auto-installation. But it solves the problem partly. See the conditions mentioned above.
This problem is just one of the consequences of a more general problem which is described in https://github.com/eclipse/che/issues/22338.
I believe it makes little sense to consider the current issue as an independent one. We can close this issue as it's not relevant anymore for the new installations. Only for the upgraded from 7.71 and earlier and in a specific case only (mentioned above).

If we solve https://github.com/eclipse/che/issues/22338, all the derivative issues will gone. In addition to the current issue, it includes those that are related to displaying the wrong recently accessed VS Code objects in a newly running Workdpace, like https://github.com/eclipse/che/issues/21813

**Below are the technical details which should allow better understand why fixing https://github.com/eclipse/che/issues/22338 is the right way to solve the current issue with the minimal efforts**
**By default,** VS Code's built-in logic gathers the information about the extensions recommendations from different sources. When it detects that there're recommended extensions should be installed, it asks the user what's their choice.
**On Che-Code side,** we implemented the recommendations auto-installation feature in a very simple way. We just [simulating clicking](https://github.com/che-incubator/che-code/blob/74caf049651391dc157b25747237a8c67d71f10d/code/src/vs/workbench/contrib/extensions/browser/extensionRecommendationNotificationService.ts#L271-L274) the `Install` button in the following notification (before it actually appears): 

<img width="455" alt="image" src="https://github.com/eclipse/che/assets/1636395/0be52ca6-2c89-4ab3-8ff6-fb5dfe070ae0">

**Before** [VS Code's upstream patch](https://github.com/microsoft/vscode/pull/186983/files), clicking `Install` button stored special setting to the browser's IndexedDB - `extensionsAssistant/workspaceRecommendationsIgnore`. The idea was the following: once the user clicked `Install`, VS Code should not check the recommendations on the subsequent run of the same workspace. 
**After** the VS Code's upstream patch, `extensionsAssistant/workspaceRecommendationsIgnore` is stored only when the user is explicitly clicked `Don't show again`:

<img width="455" alt="image" src="https://github.com/eclipse/che/assets/1636395/8b1128e1-2501-4a42-a436-d093a92bd8b1">

The problem here is that Che-Code do not distinguish two different Che Workspaces with same name. So, the same IndexedDB is used for all new Workspaces that re-use the name of any previously existed Workspace on the same Che instance.

There was an attempt to fix this problem in https://github.com/che-incubator/che-code/pull/148/files. It worked well as long as DevWorkspace ID was part of the page URL. Once we switched the default URL generating to be Workspace name-based, we've got this problem again. As the page URL is not unique anymore.

So, to fix all these issues we need to take into account a DevWorkspace ID when generating the VS Code' Workspace ID in https://github.com/che-incubator/che-code/pull/148/files

cc @l0rd @vitaliy-guliy

**Comment by azatsarynnyy** [comment]
I can reproduce the same behavior on Codespaces with https://github.com/che-samples/quarkus-quickstarts.

**Comment by azatsarynnyy** [comment]
A related upstream issue was opened recently https://github.com/microsoft/vscode/issues/192253

**Comment by azatsarynnyy** [comment]
Fixed by https://github.com/eclipse/che/issues/22338
